Adds an IR builder

kaushikcfd opened the pull request:
- This branch will have a cleaner diff with the one in #8.
- Although all the operations haven't been exposed through the builder yet, I thought it would be better to get the fundamental pieces in first.
- An example of building a saxpy using this builder:
```python
builder = IRBuilder()
F64 = builder.F64
Mref1D = builder.MemRefType(shape=(None, ), dtype=F64)

mlirfile = builder.make_mlir_file()
module = mlirfile.module

with builder.goto_block(builder.make_block(module.region)):
    saxpy_fn = builder.function("saxpy")

block = builder.make_block(saxpy_fn.region)
builder.position_at_entry(block)

a, x, y = builder.add_function_args(saxpy_fn, [F64, Mref1D, Mref1D])
c0 = builder.index_constant(0)
n = builder.dim(x, c0, builder.INDEX)

f = builder.affine_for(0, n)
i = f.index

with builder.goto_block(builder.make_block(f.region)):
    axi = builder.mulf(builder.affine_load(x, i, Mref1D), a, F64)
    axpyi = builder.addf(builder.affine_load(y, i, Mref1D), axi, F64)
    builder.affine_store(axpyi, y, i, Mref1D)

builder.ret()
```

kaushikcfd commented:
Low volume ping: Any suggestions here?

tbennun commented:
@kaushikcfd apologies, there are a lot of deadlines right now so I haven't had time to look into this yet. Thank you for sticking around though! Should be over in the next couple of weeks.

kaushikcfd commented:
@tbennun: No worries. Good luck!

tbennun commented on tests/test_builder.py:
```diff
@@ -0,0 +1,111 @@
+    c0 = builder.index_constant(0)
+    n = builder.dim(x, c0, builder.INDEX)
+
+    f = builder.affine_for(0, n)
```
General question about dialects and the IR builder: shouldn't dialects somehow extend the general builder? right now it looks like new dialects would all add to the `IRBuilder` class.

kaushikcfd replied:
My aim was to provide a builder for the core-dialects. I could have had a builder for each of the core dialect, but I figured it isn't buying us anything.

A user could just derive this and add ops correpsonding to their custom dialect.

tbennun replied:
shouldn't the default dialects serve as an example on how to extend then, i.e., with `builder.affine.for`? As opposed to numba or other IR builders, MLIR revolves around its extensible dialects, so the API should somewhat match. I'd propose one can construct an IR builder as follows:
```python
builder = IRBuilder(affine=AffineIRBuilder(), ...) # for other dialects
```
then, the individual dialects can be used as `builder.affine.load(...)` or `builder.linalg.dot`.

kaushikcfd replied:
I quite liked this idea of namespacing the dialect ops. Implemented, thanks a lot!

tbennun reviewed (commented):
Nice API! A few comments, a question, and of course this is also blocked by #8

kaushikcfd commented on mlir/builder/__init__.py:
```diff
@@ -0,0 +1,512 @@
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
```
Sure, removed!

kaushikcfd commented on setup.py:
```diff
@@ -25,7 +25,7 @@
     include_package_data=True,
     install_requires=[
-        'lark-parser', 'parse'
+        'lark-parser', 'parse', 'pytools',
```
Agreed, put a boiled-down version of his implementation into `mlir/builder/ung.py` and avoided a new dep.

kaushikcfd commented on mlir/builder/__init__.py:
```diff
@@ -0,0 +1,512 @@
+        self.position = len(block.body)
+
+    @contextmanager
+    def goto_block(self, block: ast.Block):
```
We should blame the numba folks for [this :)](https://llvmlite.readthedocs.io/en/latest/user-guide/ir/ir-builder.html?highlight=goto_block#llvmlite.ir.IRBuilder.goto_block).  Given it's slightly popular in the community, I was reluctant to invent a new name for the identical operation.

kaushikcfd commented on doc/conf.py:
```diff
@@ -9,6 +9,5 @@
 templates_path = ['_templates']
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
 html_theme = 'sphinx_rtd_theme'
-html_static_path = ['_static']
```
sphinx would give a warning `_static` directory not found. I guess it's not a big deal. Reverted this change.

kaushikcfd commented on mlir/builder/__init__.py:
```diff
@@ -0,0 +1,512 @@
+import mlir.dialects.standard as std
+import mlir.dialects.affine as affine
+from typing import Optional, Tuple, Union, List
+from pytools import UniqueNameGenerator
```
Agreed, simply copy-pasted a boiled-down version of his UniqueNameGenerator in `mlir/builder/ung.py`. (Note: had to attach the corresponding MIT license for the copied piece of code to avoid violating the license's usage policies.)

kaushikcfd commented on mlir/builder/__init__.py:
```diff
@@ -0,0 +1,512 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import mlir.astnodes as ast
```
Done, thanks!

kaushikcfd commented:
@tbennun: Thanks for taking a look, this is ready! I have rebased this branch onto master and also addressed your queries/suggestions. I have arranged the commits in an order that might help reviewing the patch in a sequential order.

tbennun commented on mlir/builder/ung.py:
```diff
@@ -0,0 +1,81 @@
+"""A boiled down version of UniqueNameGenerator from inducer/pytools"""
```
Please add a link to the GitHub repo as well.

kaushikcfd replied:
Done!

tbennun reviewed (changes requested):
Thanks! Another round of comments, this time a bit more fundamental w.r.t. dialects and the builder API.

kaushikcfd commented on mlir/builder/__init__.py:
```diff
@@ -0,0 +1,501 @@
+"""
+
+
+class IRBuilder:
```
TIL, thanks. Fixed!

kaushikcfd commented on mlir/astnodes.py:
```diff
@@ -724,22 +724,56 @@ def dump(self, indent: int = 0) -> str:
+    # TODO: The operands in these expressions should be wrapped in
+    # "AffineParens" to enforce operator precedence whenever it might be
+    # violated. Since it is only supposed to cater MLIR AST having
+    # "Parentheses" node makes sense.
```
Consider the code below
```python
In [1]: import mlir.astnodes as mast

In [2]: a = mast.AffineSsa("a")

In [3]: b = mast.AffineSsa("b")

In [4]: c = mast.AffineSsa("c")

In [5]: ((a + b + c) * 2).dump()
Out[5]: '(((%a + %b) + %c) * 2)'
```
Notice the unnecessary parentheses in the above code, There are 2 ways out of it:
1. Make the `dump` logic smarter, or,
2. Make the __add__, __mul__ methods smarter about emitting the AffineParens

I figured in the future, when this starts troubling us enough we should go with (2).

Clarified the comment in the latest version describing this.

tbennun replied:
@kaushikcfd Just noting that this happens in other "unparser"s as well:
```python
>>> import ast
>>> import astunparse
>>> print(astunparse.unparse(ast.parse('(a + b + c) * 2')))

(((a + b) + c) * 2)
```

It's mostly a matter of `dump` logic AFAIK.

tbennun reviewed (approved):
This looks much better.
